kotest-property/src/commonMain/kotlin/io/kotest/property/state/state.kt

```diff
@@ -0,0 +1,54 @@
+package io.kotest.property.state
+
+import io.kotest.property.RandomSource
+
+/**
+ * A function that is applied against the current state to mutate and/or test.
+ */
+interface Action<T> {
+
+   /**
+    * A predicate which decides if this action should be applied.
+    *
+    * For example, if you were writing a stateful test for a stack, you would want
+    * to skip 'pop' operations if the stack was empty.
+    *
+    * @return true if this action should be applied or false if the action should be skipped.
+    */
+   suspend fun precondition(state: T): Boolean
+
+   /**
+    * Invokes the next transition to the state.
+    *
+    * This function should contain your test assertions.
+    */
+   suspend fun run(state: T, rs: RandomSource)
+
+   companion object {
+
+      /**
+       * Creates a new [Action] that is always applied (precondition is set to constant true).
+       */
+      operator fun <T> invoke(f: (T, RandomSource) -> Unit): Action<T> {
+         return object : Action<T> {
+            override suspend fun precondition(state: T): Boolean = true
+            override suspend fun run(state: T, rs: RandomSource) = f(state, rs)
+         }
+      }
+   }
+}
+
+suspend fun <T> checkState(input: T, vararg actions: Action<T>) {
+   checkState(input, 1000, *actions)
+}
+
+suspend fun <T> checkState(input: T, iterations: Int, vararg actions: Action<T>) {
+   checkState(input, iterations, RandomSource.default(), *actions)
+}
+
+suspend fun <T> checkState(input: T, iterations: Int, rs: RandomSource, vararg actions: Action<T>) {
+   repeat(iterations) {
+      val action = actions.random()
```

Comment:
Sam, is this intended? as a user this will randomize my action order.. which I would consider a bug..?

Reply:
You just provide the actions, and Kotest randomly makes a sequence of them. That's the point here, you might get ABCDBDBCBCAAABCBABADBDABDADA so it's like property testing but with actions/state transitions.

Reply:
right, say we have `checkState(in, ..., action1, action2, action3)` what do I expect at the end of each test? As a user, do I:
- have only one of the action executed, or
- have these actions executed in sequence, or
- have these actions executed in parallel, or
- have these actions executed in random sequence

Reply:
executed in a random sequence of `iterations`.

Reply:
right, ok - yeah that makes sense. I guess in future improvements, we can have some sort of setting which allow user to select the behaviours, maybe using an enumerable, perhaps something like:
```kotlin
enum class ActionExecutionMode { Random, Sequential, Parallel }
```

We do have a couple of tests for race-conditions to simulate a parallel processing of many out of order events.

Reply:
Having said that, shouldn't this be `actions.random(rs.random)`?

Reply:
I like the enum idea, and yes we should be using rs.random.